pyigm/metallicity/mcmc.py

```diff
@@ -0,0 +1,887 @@
+        self.threads=threads
+        self.info=infodata
+        self.outsave=outsave
+        self.paramguess=0
+        self.final=0
+        self.optim=optim      #flag to switch type of optimisation
+        self.effnhi=effnhi    #flag to use effective NHI
+
+        #load the grid of models to memory
+        self.loadmodel(data,model)
+
+        #At this point self.data, self.mod_ions, self.colm contain only what you need
+        #Initialise the interpolator next
+
+    def loadmodel(self, data, model):
+        """ Utility function that loads the model
+
+        Parameters
+        ----------
+        data
+        model
+
+        Returns
+        -------
+
+        """
+
+
+        #load the model in a format that can be handled later on
+        fil=open(model)
+        modl=pickle.load(fil)
+        fil.close()
+
+        #unpack axis tag, axis value, grid column, grid ions
+        self.mod_axistag=modl[0]
+        self.mod_axisval=[]
+
+        #define the dimension of the problem
+        self.ndim=len(self.mod_axistag)
+        self.nmodels=1
+        for tt in self.mod_axistag:
+            self.nmodels=self.nmodels*(modl[1][tt]).size
+            #append axis value in a list
+            self.mod_axisval.append(modl[1][tt])
+
+        print("The problem as dimension {}".format(self.ndim))
+        print("There are {} models".format(self.nmodels))
+
+
+        #now queue up the model and data
+        self.nions=0
+        self.data=[]          #obs with corresponding ions in model
+        self.mod_colm=[]      #columns for the ions in model matched to obs
+        self.mod_colm_tag=[]  #ions in obs & model
+
+        #loop over all the ions observed
+        for obs in data:
+
+            #check obs one by one for corresponding entries in model
+            if modl[2].has_key(obs[0]):
+
+                #stack the observable
+                #[list of tuples for observations with (ion,column,error)]
+                self.data.append(obs)
+                #stack the columns for the models
+                #[list of ndim arrays of columns for each ion]
+                self.mod_colm.append(modl[3]['N_'+obs[0]])
+                self.mod_colm_tag.append(obs[0])
+                self.nions=self.nions+1
+
+            else:
+
+                print("{} not in the grid: skip this ion!".format(obs[0]))
+
+        print("Handling {} ions".format(self.nions))
+
+        #at last store the NHI grid if useful for effective NHI
+        if(self.effnhi):
+            print('Using effective NHI.. Load values')
+            self.nhigrid=modl[3]['N_HI']
+
+        return
+
+    def plotinfo(self, sampler):
+        """ Function that makes some useful plots of stuff
+
+        Parameters
+        ----------
+        sampler
+
+        Returns
+        -------
+
+        """
+
+        print('Writing outputs...')
+
+        #pickle the results to disk
+        wout=open(self.outsave+'/'+self.info['name']+'_emcee.pkl','w')
```

Comment:
I'm willing to do hd5.  I'm not keen on Python specific files, just
like I never liked using IDL save files.  Am happy to defer to
your wisdom here.  I presume these may get very large if 
saved as ASCII (i.e. JSON).  So a binary format makes sense.
But it might be that a mix of the two is best, i.e. a JSON file for
the key parameters and an HDF5 file for the PDF output.  Thoughts?